backtester/data/data.go

```diff
@@ -51,84 +57,132 @@ func (h *HandlerPerCurrency) GetDataForCurrency(ev common.EventHandler) (Handler
 
 // Reset returns the struct to defaults
 func (h *HandlerPerCurrency) Reset() {
+	if h == nil {
+		return
+	}
 	h.data = nil
 }
 
 // Reset loaded data to blank state
 func (b *Base) Reset() {
+	if b == nil {
+		return
+	}
 	b.latest = nil
 	b.offset = 0
 	b.stream = nil
 }
 
 // GetStream will return entire data list
-func (b *Base) GetStream() []common.DataEventHandler {
+func (b *Base) GetStream() []Event {
 	return b.stream
 }
 
 // Offset returns the current iteration of candle data the backtester is assessing
-func (b *Base) Offset() int {
+func (b *Base) Offset() int64 {
 	return b.offset
 }
 
 // SetStream sets the data stream for candle analysis
-func (b *Base) SetStream(s []common.DataEventHandler) {
+func (b *Base) SetStream(s []Event) {
 	b.stream = s
+	// due to the Next() function, we cannot take
+	// stream offsets as is, and we re-set them
+	for i := range b.stream {
+		b.stream[i].SetOffset(int64(i + 1))
+	}
 }
 
 // AppendStream appends new datas onto the stream, however, will not
 // add duplicates. Used for live analysis
-func (b *Base) AppendStream(s ...common.DataEventHandler) {
-	for i := range s {
-		if s[i] == nil {
+func (b *Base) AppendStream(s ...Event) {
+candles:
+	for x := range s {
+		if s[x] == nil || !b.equalSource(s[x]) {
 			continue
 		}
-		b.stream = append(b.stream, s[i])
+		for y := range b.stream {
```

Comment:
I've added a todo, I think it would be good. But as you're going to see, there's been quite a few changes as a result of these comments :smile: this kind of efficiency improvement can be made in its own PR